From 19df7a550772f845c1b690b9d8f4c9eb1050a3bc Mon Sep 17 00:00:00 2001
From: Xufeng Zhang <xufeng.zhang@windriver.com>
Date: Mon, 4 Nov 2013 15:41:21 +0800
Subject: [PATCH] ARM: OMAP3: Move sr1_hwmod and sr2_hwmod to chip specific hwmod

Backport based on:
upstream commit d6504acd2125984c61dce24727dd3842d0144015
("OMAP2+: hwmod: remove OMAP_CHIP*")
upstream commit 7e89098cd63a188932043258d54688e965750e2c
("ARM: OMAP: AM35x: remove hwmods that aren't generic")
upstream commit ace9021698ef7d298e6d184ae2880b1cb3ebc4c7
("ARM: OMAP3: hwmod: fix variant registration and remove SmartReflex from common list")

Move sr1_hwmod and sr2_hwmod from generic hwmod part(omap3xxx_hwmods)
to chip specific part(omap34xx_hwmods and omap36xx_hwmods).
It also add a missed flag(CHIP_IS_OMAP3430ES3_1) to omap34xx_sr1_hwmod.

Signed-off-by: Xufeng Zhang <xufeng.zhang@windriver.com>
---
 arch/arm/mach-omap2/omap_hwmod_3xxx_data.c |   46 ++++++++++++++++++++++++++--
 1 files changed, 43 insertions(+), 3 deletions(-)

diff --git a/arch/arm/mach-omap2/omap_hwmod_3xxx_data.c b/arch/arm/mach-omap2/omap_hwmod_3xxx_data.c
index 52e56c7..14974b5 100644
--- a/arch/arm/mach-omap2/omap_hwmod_3xxx_data.c
+++ b/arch/arm/mach-omap2/omap_hwmod_3xxx_data.c
@@ -547,7 +547,8 @@ static struct omap_hwmod omap34xx_sr1_hwmod = {
 	.slaves		= omap3_sr1_slaves,
 	.slaves_cnt	= ARRAY_SIZE(omap3_sr1_slaves),
 	.omap_chip	= OMAP_CHIP_INIT(CHIP_IS_OMAP3430ES2 |
-			CHIP_IS_OMAP3430ES3_0),
+			CHIP_IS_OMAP3430ES3_0 |
+			CHIP_IS_OMAP3430ES3_1),
 	.flags		= HWMOD_SET_DEFAULT_CLOCKACT,
 };
 
@@ -627,14 +628,53 @@ static __initdata struct omap_hwmod *omap3xxx_hwmods[] = {
 	&omap3xxx_uart2_hwmod,
 	&omap3xxx_uart3_hwmod,
 	&omap3xxx_uart4_hwmod,
+	NULL,
+};
+
+/* 34xx-only hwmods (all ES revisions) */
+static __initdata struct omap_hwmod *omap34xx_hwmods[] = {
 	&omap34xx_sr1_hwmod,
 	&omap34xx_sr2_hwmod,
+	NULL
+};
+
+/* 36xx-only hwmods (all ES revisions) */
+static __initdata struct omap_hwmod *omap36xx_hwmods[] = {
 	&omap36xx_sr1_hwmod,
 	&omap36xx_sr2_hwmod,
-	NULL,
+	NULL
 };
 
 int __init omap3xxx_hwmod_init(void)
 {
-	return omap_hwmod_init(omap3xxx_hwmods);
+	int r;
+	struct omap_hwmod **h = NULL;
+	unsigned int rev;
+
+	/* Register hwmods common to all OMAP3 */
+	r = omap_hwmod_init(omap3xxx_hwmods);
+	if (r < 0)
+		return r;
+
+	rev = omap_rev();
+
+	/*
+	 * Register hwmods common to individual OMAP3 families, all
+	 * silicon revisions (e.g., 34xx, or AM3505/3517, or 36xx)
+	 * All possible revisions should be included in this conditional.
+	 */
+	if (rev == OMAP3430_REV_ES1_0 || rev == OMAP3430_REV_ES2_0 ||
+	    rev == OMAP3430_REV_ES2_1 || rev == OMAP3430_REV_ES3_0 ||
+	    rev == OMAP3430_REV_ES3_1 || rev == OMAP3430_REV_ES3_1_2) {
+		h = omap34xx_hwmods;
+	} else if (rev == OMAP3630_REV_ES1_0 || rev == OMAP3630_REV_ES1_1 ||
+		   rev == OMAP3630_REV_ES1_2) {
+		h = omap36xx_hwmods;
+	} else {
+		return r;
+	};
+
+	r = omap_hwmod_init(h);
+	
+	return r;
 }
-- 
1.7.0

